worker/caasrbacmapper/mapper.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package caasrbacmapper
+
+import (
+	"sync"
+	"time"
+
+	"github.com/juju/errors"
+	"gopkg.in/juju/worker.v1"
+	"gopkg.in/juju/worker.v1/catacomb"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	core "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/juju/juju/caas/kubernetes/provider"
+)
+
+// Mapper describes an interface for mapping k8s service account UID's to juju
+// application names.
+type Mapper interface {
+	// AppNameForServiceAccount fetches the juju application name associated
+	// with a given kubernetes service account UID. If no result is found
+	// errors.NotFound is returned. All other errors should be considered
+	// internal to the interface operation.
+	AppNameForServiceAccount(types.UID) (string, error)
+}
+
+// MapperWorker is a Mapper that also implements the worker interface
+type MapperWorker interface {
+	Mapper
+	worker.Worker
+}
+
+// DefaultMapper is a default implementation of the MapperWorker interface. It's
+// responsible for watching ServiceAccounts on a given model and
+type DefaultMapper struct {
+	catacomb     catacomb.Catacomb
+	lock         *sync.RWMutex
+	logger       Logger
+	saInformer   core.ServiceAccountInformer
+	saNameUIDMap map[string]types.UID
+	saUIDAppMap  map[types.UID]string
+	workQueue    workqueue.RateLimitingInterface
+}
+
+// AppNameForServiceAccount implements Mapper interface
+func (d *DefaultMapper) AppNameForServiceAccount(id types.UID) (string, error) {
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+	appName, found := d.saUIDAppMap[id]
+	if !found {
+		return "", errors.NotFoundf("no service account for app found with id %v", id)
+	}
+	return appName, nil
+}
+
+func (d *DefaultMapper) enqueueServiceAccount(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		d.logger.Errorf("failed enqueuing service account: %v", err)
+		return
+	}
+	d.workQueue.Add(key)
+}
+
+// Kill implements Worker interface
+func (d *DefaultMapper) Kill() {
+	d.catacomb.Kill(nil)
+}
+
+func (d *DefaultMapper) loop() error {
+	defer d.workQueue.ShutDown()
+	d.saInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Comment:
probably to add the handler in  constructor `NewMapper`

Reply:
Done & Done